Speed up HistogramObserver by vectorizing critical path

durumu opened the pull request:
Roughly a 22x speedup over the code this replaces when tested on ResNet18 on a devvm using CPU only, using default parameters for HistogramObserver (i.e. 2048 bins). The script I ran to test this is [here](https://gist.github.com/durumu/590803fadd171da23338ee769cf2de9c). 

Roughly a 14x speedup when tested using the benchmark from #42138 (also CPU only). 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41041 Speed up HistogramObserver by vectorizing critical path**
* #42138 Add benchmark for calculate_qparams

Differential Revision: [D22400755](https://our.internmc.facebook.com/intern/diff/D22400755)

durumu committed:
Speed up HistogramObserver by vectorizing critical path

[ghstack-poisoned]

vkuzo commented:
nice!  In general looks good, can we just add to the test plan:
* how did we measure performance
* how did we verify numerical correctness

durumu committed:
Update on "Speed up HistogramObserver by vectorizing critical path"


Differential Revision: [D22400755](https://our.internmc.facebook.com/intern/diff/D22400755)

[ghstack-poisoned]

mruberry commented on torch/quantization/observer.py:
```diff
@@ -682,50 +681,44 @@ def _compute_quantization_error(next_start_bin, next_end_bin, norm_type):
-                    norm = norm + _get_norm(delta_begin, delta_end, density, norm_type)
-            return norm
+
+            src_bin = torch.arange(self.bins).numpy()
```
PyTorch doesn't have a NumPy dependency for its functionality (although we do for some tests), and we shouldn't use NumPy functionality in lieu of our own. Uses of NumPy should be restricted to testing and NumPy interop.

durumu replied:
Thanks for the feedback -- I changed my code to get rid of the numpy dependency.

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -682,50 +681,44 @@ def _compute_quantization_error(next_start_bin, next_end_bin, norm_type):
-                    delta_end = src_bin_end - dst_bin_of_end_center
-                    norm = norm + _get_norm(delta_begin, delta_end, density, norm_type)
-            return norm
+
```
This can be optimized further by the following approximation: 
 Quantization error = (StepSize^2/12)*Q + sum(P[i]*(BinCenter[i]-next_start_bin)^2) + sum(Pi]*(BinCenter[i] - end_start_bin)^2). 
Q = sum(hist[next_start_bin:next_end_bin])

where the first sum is over the bins less than the start_bin and the second sum is over bins greater than the end bin. In this approximation, we only need to compute two indices: Where do the next_start_bin and next_end_bin map to in terms of the original histogram indices

durumu committed:
Update on "Speed up HistogramObserver by vectorizing critical path"


Differential Revision: [D22400755](https://our.internmc.facebook.com/intern/diff/D22400755)

[ghstack-poisoned]

durumu committed:
Update on "Speed up HistogramObserver by vectorizing critical path"


Differential Revision: [D22400755](https://our.internmc.facebook.com/intern/diff/D22400755)

[ghstack-poisoned]

durumu committed:
Update on "Speed up HistogramObserver by vectorizing critical path"


Differential Revision: [D22400755](https://our.internmc.facebook.com/intern/diff/D22400755)

[ghstack-poisoned]

durumu committed:
Update on "Speed up HistogramObserver by vectorizing critical path"


Differential Revision: [D22400755](https://our.internmc.facebook.com/intern/diff/D22400755)

[ghstack-poisoned]
